### PR TITLE
Fix #9227: Do not suspend REPL compilation unit

### LINKED ISSUE #9227: Macro MatchError on EmptyTree
## Minimized code

REPL crash with a MatchError on EmptyTree. If we do a single call on the REPL

```Scala
import scala.quoted._                                                                                                                                                                                                                  
inline def myMacro[T]: Unit = ${ myMacroImpl[T] }
def myMacroImpl[T](using qctx: QuoteContext): Expr[Unit] = '{}
println(myMacro[Int])
```

I was expecting to fail compilation after calling the macro with the error
```scala
Cannot call macro method myMacroImpl defined in the same source file
  | This location contains code that was inlined from rs$line$1:4
```
Instead, it seems that an `EmptyTree` is returned during macro expansion.
I assume it tries to continue the compilation of the line with that.

## Output

```scala
Exception in thread "main" scala.MatchError: EmptyTree (of class dotty.tools.dotc.ast.Trees$EmptyTree)
	at dotty.tools.repl.CollectTopLevelImports.topLevelImports$1(CollectTopLevelImports.scala:23)
	at dotty.tools.repl.CollectTopLevelImports.run(CollectTopLevelImports.scala:28)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:318)
	at scala.collection.immutable.List.map(List.scala:246)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:319)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:180)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:190)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:198)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:64)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:205)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:147)
	at dotty.tools.repl.ReplCompiler.runCompilationUnit(ReplCompiler.scala:156)
	at dotty.tools.repl.ReplCompiler.compile(ReplCompiler.scala:166)
	at dotty.tools.repl.ReplDriver.compile(ReplDriver.scala:231)
	at dotty.tools.repl.ReplDriver.interpret(ReplDriver.scala:195)
	at dotty.tools.repl.ReplDriver.loop$1(ReplDriver.scala:128)
	at dotty.tools.repl.ReplDriver.runUntilQuit$$anonfun$1(ReplDriver.scala:131)
	at dotty.tools.repl.ReplDriver.withRedirectedOutput(ReplDriver.scala:150)
	at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:131)
	at dotty.tools.repl.Main$.main(Main.scala:6)
	at dotty.tools.repl.Main.main(Main.scala)
[error] Nonzero exit code returned from runner: 1
[error] (dotty-compiler / Compile / runMain) Nonzero exit code returned from runner: 1
```

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/CompilationUnit.scala [hunk 1]
```diff
@@ -25,6 +26,8 @@ class CompilationUnit protected (val source: SourceFile) {
 
   def isJava: Boolean = source.file.name.endsWith(".java")
 
+  def isREPL: Boolean = source.file.name.startsWith(str.REPL_SESSION_LINE)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/CompilationUnit.scala:29]: Is it possible to generalize this a little, like `var isSuspendable = true`, and mutate the field for REPL compilation units?

I was thinking that maybe worksheets may have the same need.

File: compiler/src/dotty/tools/dotc/CompilationUnit.scala [hunk 2]
```diff
@@ -47,7 +52,14 @@ class CompilationUnit protected (val source: SourceFile) {
 
   var suspended: Boolean = false
 
+  /** Can this compilation unit be suspended */
+  def isSuspendable: Boolean = !isREPL
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/CompilationUnit.scala:56]: Indeed, I could remove `isREPL`

File: compiler/src/dotty/tools/repl/ReplCompiler.scala [hunk 3]
```diff
@@ -146,7 +146,7 @@ class ReplCompiler extends Compiler {
     val objectTermName = ctx.source.file.toString.toTermName
     objectNames.update(defs.state.objectIndex, objectTermName)
 
-    val unit = CompilationUnit(ctx.source)
+    val unit = new ReplCompilationUnit(ctx.source)
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/repl/ReplCompiler.scala:149]: Nit: I would use an anonymous class
```suggestion
    val unit = new CompilationUnit(ctx.source) {
      override def isSuspendable: Boolean = false
    }
```

Comment (nicolasstucki) [reply compiler/src/dotty/tools/repl/ReplCompiler.scala:149]: I wanted to make it possible to identify a REPL compilation unit simply. I remember seeing places where we identify them by the name of the file which we could simplify with this scheme.

**Pull request merged**